### DISCUSSION TITLE
Noun-verb instead of Verb-noun ?

### DISCUSSION THREAD

**Discussion by ghost** [PowerShell/PowerShell#17765] [Enhancements]
Hi, any feedback on this idea ? Did some research but not many results.

I Think Powershell should have used noun-verb instead of verb-noun because of autocompletion.

If you tab complete "get-" you'll have to loop around hundreds of cmdlets, instead if noun was first you'll autocomplete "Aduser-" and only get a few results like "ADuser-get", "ADuser-Set" and so on.

Maybe it would be possible to autmagically do this in a future release ? Maybe like creating aliases for each command reversing the noun and verb ?

any thought ?

**Comment by Northman-de** [comment]
*ADUser <TAB> will cycle through get-aduser, set-aduser, new-aduser
*ADUser CTRL+SPACE show you a list of the available cmdlets

And there is also the famous "get-command *aduser*".

I do not know how you work, but I know upfront whether I need get- or set- before typing a command ;-)

**Reply by ghost** [reply]
> *ADUser CTRL+SPACE show you a list of the available cmdlets

Didn't know this one, thanks :-)

> I do not know how you work, but I know upfront whether I need get- or set- before typing a command ;-)

Sure me too, my point was more about putting the less frequent term in first to limit the possible results the early possible.

**Reply by fMichaleczek** [reply]
@Northman-de 

thank you so much ❤
It will change my life forever !

**Comment by kilasuit** [comment]
Whilst I can see why this may be something that would be of use, it just will not happen & was designed this way to follow logical processing in the right way and not taking shortcuts in the end user experience,
You have to remember that the computer can only do what the user tells it to do, so logically it should, get something & then move on to the next step which may be a change/delete task. 

A lot of confusion when learning PowerShell is caused by thinking not in this way is down to us wanting to jump ahead a few steps, when we really should process each step as required one by one and not jump ahead of ourselves. Once we learn the fundamentals we can then start jumping ahead if we feel it appropriate to do so.

There are already a number of additional helper mechanisms as pointed out prior that make this easier to find and work with, once you understand the basis of stringing together a series of potentially complex tasks. In my years of teaching PowerShell to newcomers I see the same things, and that is that most want to rush ahead instead of learning the core foundations that'll really help you get going in future.

Hope the above has helped explain the background & I'd be more than happy to explain anything that you may want to take a little further via one of the other ways of engaging with the community, whether that be via slack or discord or via user groups or conferences too.

**Reply by ghost** [reply]
You are more than correct from a logical standpoint, I think I'm correct too from a usability standpoint but in the end I think logic should win over usability (at least here)

Anyway it's a very small detail in the grand scheme of Powershell things, thanks for your input.

**Comment by doctordns** [comment]
An interesting suggestgion. My take is tthat in most languages, the verb comes before the noun in sentences, a style PowerShell mimics. So "Get the AD Users" is, in PowerShell, `Get-ADUser`. You would not say  in many languages "AD users, get". Unless you are german speaking, where the verb tends to be at the end of the sentence.  But that is a miniority. 

I also find the Verb-Noun approach is easier to teach. finding/discovering/retreiving - it's all Get. Then you say what you want to get and any parameters to change the scope of the command. To me, the other way around (Noun-Verb) is just not logical.  But I guess mileage varies.

That said, I somehow doubt this will ever happen.  If this were a really important challenge for you, then you could set up some aliases in your profile, something like this:

```powershell
$Commands = Get-Command -CommandType Cmdlet 
Foreach ($Cmd in $Commands) {
  $Alias  = "$($Cmd.Noun)" + "-" + "$($Cmd.verb)"
  $Cmdlet = "$($Cmd.Verb)" + "-" + "$($Cmd.noun)"   
  Set-Alias $Alias $Cmdlet -Verbose
}
```

And if you wanted to be perverse you could create aliases for each cmdlet that does nothing! ;-)

As for the suggestion itself, I think the time for this discussion was 20-odd years ago.  Making such a change would be a major breaking change that I can not believe anyone would really want to do.  It would require every existing script to be rewritten. 

My inclination is to close this as "by design".

**Comment by ghost** [comment]
As I said it's not a challenge at all, it was just a  small idea of mine
and was really pretty much only about faster autocompletion.

I agree with your arguments on why in the end it's probably better as it is
right now.

Ho and thanks for the code snippet which is another proof that if something
in powershell doesn't suit someone one can almost always tweak it to his
liking. Damn I love this Language so much :-)

cheers.


Le jeu. 28 juil. 2022 à 17:14, Thomas Lee ***@***.***> a
écrit :
